csharp/ql/src/semmle/code/csharp/dataflow/internal/FlowSummaryImpl.qll

```diff
@@ -131,340 +178,497 @@ module Public {
       or
-      exists(SummaryOutput output, ContentListRev l |
-        this = TSummaryInternalNodeStoreState(output, l) and
-        result = "output: " + output + ", to store: " + l
+      exists(SummaryComponentStack s |
+        this = TSummaryNodeOutputState(s) and
+        result = "to write: " + s
       )
     }
   }
 
   /**
-   * Holds if an internal summary node is needed for the state `state` in summarized
+   * Holds if `state` represents having read the `i`th argument for `c`. In this case
+   * we are not synthesizing a data-flow node, but instead assume that a relevant
+   * parameter node already exists.
+   */
+  private predicate parameterReadState(SummarizedCallable c, SummaryNodeState state, int i) {
+    state.isInputState(c, SummaryComponentStack::argument(i))
+  }
+
+  /**
+   * Holds if a synthesized summary node is needed for the state `state` in summarized
    * callable `c`.
    */
-  predicate internalNodeRange(SummarizedCallable c, SummaryInternalNodeState state) {
-    state.isReadState(c, _, _, _)
+  predicate summaryNodeRange(SummarizedCallable c, SummaryNodeState state) {
+    state.isInputState(c, _) and
+    not parameterReadState(c, state, _)
     or
-    state.isStoreState(c, _, _, _)
+    state.isOutputState(c, _)
   }
 
   pragma[noinline]
-  private Node internalNodeLastReadState(
-    SummarizedCallable c, SummaryInput input, ContentList inputContents
-  ) {
-    exists(SummaryInternalNodeState state |
-      state.isReadState(c, input, inputContents, inputContents.length()) and
-      result = internalNode(c, state)
+  private Node summaryNodeInputState(SummarizedCallable c, SummaryComponentStack s) {
+    exists(SummaryNodeState state | state.isInputState(c, s) |
+      result = summaryNode(c, state)
+      or
+      exists(int i |
+        parameterReadState(c, state, i) and
+        result.(ParameterNode).isParameterOf(c, i)
+      )
     )
   }
 
   pragma[noinline]
-  private Node internalNodeFirstStoreState(
-    SummarizedCallable c, SummaryOutput output, ContentList outputContents
-  ) {
-    exists(SummaryInternalNodeState state |
-      state.isStoreState(c, output, outputContents, outputContents.length()) and
-      result = internalNode(c, state)
+  private Node summaryNodeOutputState(SummarizedCallable c, SummaryComponentStack s) {
+    exists(SummaryNodeState state |
+      state.isOutputState(c, s) and
+      result = summaryNode(c, state)
     )
   }
 
   /**
-   * Holds if there is a local step from `pred` to `succ`, which is synthesized
-   * from a flow summary.
+   * Holds if a write targets `post`, which is a post-update node for the `i`th
+   * parameter of `c`.
+   */
+  private predicate isParameterPostUpdate(Node post, SummarizedCallable c, int i) {
+    post = summaryNodeOutputState(c, SummaryComponentStack::argument(i))
+  }
+
+  /** Holds if a parameter node is required for the `i`th parameter of `c`. */
+  predicate summaryParameterNodeRange(SummarizedCallable c, int i) {
```

Comment:
Is this unused?

Reply:
I guess that's for the potential case where a language doesn't have all the necessary `ParameterNode`s by default?

Reply:
Indeed.